### PR TITLE
Client API

### LINKED ISSUE #442: Platform Client
We need a client for the Platform

### PR DESCRIPTION
**Pull request by haiodo**
Refactor to get client API

Closes #442 

Signed-off-by: Andrey Sobolev <haiodo@gmail.com>

### CODE REVIEW

File: packages/client/src/rpc.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+  async function request<R> (method: string, ...params: any[]): Promise<R> {
+    return await new Promise<any>((resolve, reject) => {
+      const id = ++lastId
+      requests.set(id, {
+        resolve,
+        reject
+      })
+        getWebSocket().then((ws: WebSocket) => { // eslint-disable-line
```
Comment (sumboid) [inline packages/client/src/rpc.ts:92]: Formatting seems to be invalid due to `eslint-disable-line`

**Review by sumboid — COMMENTED** [review]
Just multiple minor comments

File: packages/core/src/model.ts [hunk 2]
```diff
@@ -870,7 +870,26 @@ export class Model implements Storage {
...
+    const oKeys = new Set<string>(Object.keys(_attributes))
+
+    // Make a part of query with values in object.
+    let keys = 0
+    for (const oe of Object.entries(sort)) {
+      if (oKeys.has(oe[0])) {
+        keys++
+      }
+    }
+    return keys > 0
```
[the remark this answers is not in the document]
Comment (haiodo) [reply packages/core/src/model.ts:892]: Nice, Thanks.

File: packages/client/src/rpc.ts (hunk 1 above)
Comment (haiodo) [reply packages/client/src/rpc.ts:92]: Fixed.

File: packages/client/src/rpc.ts [hunk 3]
```diff
@@ -28,30 +12,23 @@ export enum EventType {
...
-
-    const token = platform.getMetadata(core.metadata.Token)
-
-    if (token === undefined) {
-      platform.broadcastEvent(PlatformStatus, new Status(Severity.ERROR, PlatformStatusCodes.AUTHENTICATON_REQUIRED, 'Authentication is required'))
-      return Promise.reject(new Error('authentication required'))
-    }
     return new Promise<WebSocket>((resolve) => {
       // Let's sure token is valid one
+      const WebSocket = (typeof window !== 'undefined' ? window.WebSocket : require('ws'))
```
Comment (z2sx) [inline packages/client/src/rpc.ts:29]: I do not like this: I believe both node and browser use cases could be properly designed to avoid such 'hacks'

Comment (haiodo) [reply packages/client/src/rpc.ts]: Added a RpcSocket to hold a implementation specific for different platforms.

Comment (z2sx) [reply packages/client/src/rpc.ts]: This is same: what is `RpcSocket` type? Can we create node WebSocket in node client and pass browser socket in plugin?

Comment (haiodo) [reply packages/client/src/rpc.ts]: they have incompatible interfaces, I've only could do so with using `any`.

Comment (z2sx) [reply packages/client/src/rpc.ts]: > they have incompatible interfaces

So current approach sucks more then :)

Comment (haiodo) [reply packages/client/src/rpc.ts]: I've found a variant I've like after some findings, it is actually possible to use package.json, "browser" field to specify a different source file for module, so for web we have variant with WebSocket under hood with types checked and for other we have 'ws' WebSocket version with all type checked calls. No hacks, but it uses intermediate RpcClient interface to unify some parts of code.

File: packages/client/package.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "@anticrm/client",
+  "version": "0.2.0",
+  "main": "src/index.ts",
+  "browser": "src/index.browser.ts",
```
Comment (haiodo) [inline packages/client/package.json:5]: This field make a way better usage for WebSockets :-)

File: packages/client/package.json [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    "eslint-plugin-promise": "^5.1.0",
+    "eslint-plugin-standard": "^5.0.0",
+    "jest": "^26.6.3",
+    "typescript": "^4.2.4"
+  },
+  "dependencies": {
+    "@anticrm/platform": "^0.2.0",
+    "@anticrm/rpc": "^0.2.0",
+    "@anticrm/foundation": "^0.2.0",
+    "ws": "^7.4.5"
```
Comment (haiodo) [inline packages/client/package.json:39]: Not sure if we possible could mark this dependency as optional?

File: packages/client/src/rpc.ts (hunk 3 above)
Comment (z2sx) [reply packages/client/src/rpc.ts]: Why do not do what you will do in statically typed language? You have to create `ws` socket from `client` package and `browser` from `platfrom-core`. So you do not need `browser` field.

Comment (haiodo) [reply packages/client/src/rpc.ts]: Get rid of browser field, and moved creation into platform-core and leave a node version inside index.ts of client.

**Review by z2sx — APPROVED** [review]

**Pull request merged**